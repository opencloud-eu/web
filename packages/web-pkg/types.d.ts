import { Ability } from '@opencloud-eu/web-client'
import {
  ArchiverService,
  ClientService,
  LoadingService,
  PreviewService,
  PasswordPolicyService
} from './src/services'
import { Route, Router } from 'vue-router'

export * from './src'

declare module 'vue' {
  interface ComponentCustomProperties {
    $ability: Ability
    $archiverService: ArchiverService
    $clientService: ClientService
    $loadingService: LoadingService
    $previewService: PreviewService
    $passwordPolicyService: PasswordPolicyService
    $router: Router
    $route: Route
  }
}
