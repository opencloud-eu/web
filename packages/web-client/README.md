# web-client

The `web-client` is a standalone package that allows you to interact with the [OpenCloud](https://github.com/opencloud-eu/opencloud/) APIs via TypeScript. It provides an abstraction layer between the server and a (web-) application that converts API data into objects with helpful types and utilities. This abstraction ensures that users of the APIs don't need in-depth knowledge about them, such as required methods or returned status codes.

The supported APIs are:

- Graph (drives, sharing, user & group management)
- OCS (capabilities & url signing)
- WebDAV (file operations)

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-client

$ pnpm add @opencloud-eu/web-client

$ yarn add @opencloud-eu/web-client
```

## Usage

### Graph

The graph client needs to be instantiated with a base URI corresponding to your OpenCloud deployment and an axios instance. The axios instance is being used for all requests, which means it needs to include all relevant headers either statically or via interceptor.

```
import axios from axios
import { graph } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const graphClient = graph(baseURI, axiosClient)
```

The following example demonstrates how to retrieve all spaces accessible to the user. A `SpaceResource` can then be used to e.g. fetch files and folders (see webdav example down below).

```
const mySpaces = await graphClient.drives.listMyDrives()
```

### OCS

The ocs client needs to be instantiated with a base URI corresponding to your OpenCloud deployment and an axios instance. The axios instance is being used for all requests, which means it needs to include all relevant headers either statically or via interceptor.

```
import axios from axios
import { ocs } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const ocsClient = ocs(baseURI, axiosClient)
```

The following examples demonstrate how to fetch capabilities and sign URLs.

```
const capabilities = await ocsClient.getCapabilities()

const signedUrl = await ocsClient.signUrl('some_url_to_sign', 'your_username')
```

### WebDav

The webdav client needs to be instantiated with a base URI corresponding to your OpenCloud deployment. You can also pass a header callback which will be called with every dav request.

```
import { webdav } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const webDavClient = webdav(
	baseURI,
	() => ({ Authorization: accessToken })
)
```

The following example demonstrates how to list all resources of a given `SpaceResource` (see above how to fetch space resources).

```
const { resource, children } = await webDavClient.listFiles(spaceResource)
```
