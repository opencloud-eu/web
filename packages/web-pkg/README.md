# web-pkg

`web-pkg` is a package that provides utilities, most importantly a variety of components and composables, that can be useful when developing apps and extensions for OpenCloud Web.

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-pkg --save-dev

$ pnpm add -D @opencloud-eu/web-pkg

$ yarn add @opencloud-eu/web-pkg --dev
```

It's recommended to install this package as a dev dependency because it's only really needed for providing autocompletion in your IDE and unit tests. In a runtime context, the OpenCloud Web runtime provides the actual implementation.
