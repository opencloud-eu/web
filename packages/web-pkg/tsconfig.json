{
  "extends": "@opencloud-eu/tsconfig",

  // necessary for building web-pkg standalone, otherwise we get errors with inferred types
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "@vue/shared": ["node_modules/@vue/shared/"],
      "@codemirror/autocomplete": ["./node_modules/@codemirror/autocomplete"],
      "vue/jsx-runtime": ["./node_modules/vue/jsx-runtime"],
      "@popperjs/core": ["./../design-system/node_modules/@popperjs/core"]
    }
  },

  // include component types of the design-system
  "include": ["../design-system/types.d.ts"]
}
