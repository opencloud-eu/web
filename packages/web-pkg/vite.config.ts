import { join, resolve } from 'path'
import { defineConfig, searchForWorkspaceRoot } from 'vite'
import dts from 'vite-plugin-dts'
import { nodePolyfills } from 'vite-plugin-node-polyfills'
import vue from '@vitejs/plugin-vue'
import pkg from './package.json'

const projectRootDir = searchForWorkspaceRoot(process.cwd())
const external = [...Object.keys(pkg.dependencies)]

export default defineConfig({
  resolve: {
    alias: {
      crypto: join(projectRootDir, 'polyfills/crypto.js')
    }
  },
  css: {
    preprocessorOptions: {
      scss: {
        additionalData: `
          @use "sass:math";
          @use "sass:string";
          @use "sass:meta";
        `,
        silenceDeprecations: ['legacy-js-api', 'import']
      }
    }
  },
  build: {
    lib: {
      entry: resolve(__dirname, 'src/index.ts'),
      name: 'web-pkg',
      fileName: 'web-pkg'
    },
    rollupOptions: {
      external: external.filter(
        (e) =>
          // something is off with this lib, see https://github.com/ahmadjoya/generate-password-lite/issues/8
          e !== 'js-generate-password'
      )
    }
  },
  plugins: [
    vue(),
    nodePolyfills({
      exclude: ['crypto']
    }),
    dts({ exclude: ['**/tests'] })
  ]
})
