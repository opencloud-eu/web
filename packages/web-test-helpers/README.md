# web-test-helpers

This packages provides utilities for unit testing within the OpenCloud app ecosystem.

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-test-helpers --save-dev

$ pnpm add -D @opencloud-eu/web-test-helpers

$ yarn add @opencloud-eu/web-test-helpers --dev
```

Make sure that you have `vue`, `@vue/test-utils`, and `@opencloud-eu/web-pkg` as dev dependencies of your app since those are peer dependencies of this package.
