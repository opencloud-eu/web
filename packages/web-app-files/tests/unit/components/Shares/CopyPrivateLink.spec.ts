import { mock } from 'vitest-mock-extended'
import { Resource } from '@opencloud-eu/web-client'
import { defaultPlugins, mount } from '@opencloud-eu/web-test-helpers'
import CopyPrivateLink from '../../../../src/components/Shares/CopyPrivateLink.vue'
import { useMessages } from '@opencloud-eu/web-pkg'
import { useClipboard } from '@vueuse/core'
import { ref } from 'vue'

const resource = mock<Resource>({
  type: 'folder',
  owner: {
    id: 'marie',
    displayName: 'Marie'
  },
  mdate: 'Wed, 21 Oct 2015 07:28:00 GMT',
  size: '740',
  name: 'lorem.txt',
  privateLink: 'https://example.com/fake-private-link'
})

vi.mock('@vueuse/core', () => ({
  useClipboard: vi.fn(() => ({
    copy: vi.fn(),
    copied: false,
    isSupported: true
  }))
}))

describe('CopyPrivateLink', () => {
  it('should render a button', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('upon clicking it should copy the private link to the clipboard button, render a success message and change icon for half a second', async () => {
    const copyMock = vi.fn()
    vi.mocked(useClipboard).mockReturnValue(
      mock<ReturnType<typeof useClipboard>>({ copy: copyMock, copied: ref(true) })
    )

    const { wrapper } = getWrapper()
    const { showMessage } = useMessages()
    expect(showMessage).not.toHaveBeenCalled()

    await wrapper.find('button').trigger('click')
    expect(copyMock).toHaveBeenCalledTimes(1)
    expect(showMessage).toHaveBeenCalledTimes(1)
  })
})

function getWrapper() {
  return {
    wrapper: mount(CopyPrivateLink, {
      props: {
        resource
      },
      global: {
        plugins: [...defaultPlugins()]
      }
    })
  }
}
