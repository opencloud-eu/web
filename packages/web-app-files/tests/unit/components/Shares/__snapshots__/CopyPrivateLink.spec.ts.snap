// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CopyPrivateLink > should render a button 1`] = `
"<div class="oc-flex oc-flex-middle copy-private-link"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-undefined oc-button-secondary oc-button-raw oc-button-secondary-raw">
    <!--v-if-->
    <!-- @slot Content of the button --> <span class="oc-icon oc-icon-s"><svg data-testid="inline-svg-stub" src="icons/file-copy-line.svg" transform-source="(svg) => {
      if (__props.accessibleLabel !== &quot;&quot;) {
        const title = document.createElement(&quot;title&quot;);
        title.setAttribute(&quot;id&quot;, svgTitleId.value);
        title.appendChild(document.createTextNode(__props.accessibleLabel));
        svg.insertBefore(title, svg.firstChild);
      };
      return svg;
    }" aria-hidden="true" focusable="false"></svg></span> <span class="oc-ml-xs">Permanent link</span>
  </button></div>"
`;
